lib/graphql-federation.module.ts

```diff
@@ -0,0 +1,204 @@
+import { GqlModuleAsyncOptions, GqlModuleOptions, GqlOptionsFactory } from './interfaces';
+import { generateString, extend, mergeDefaults, normalizeRoutePath } from './utils';
+import { GraphQLFactory } from './graphql.factory';
+
+@Module({
+  providers: [
+    GraphQLFederationFactory,
+    GraphQLFactory,
+    MetadataScanner,
+    ResolversExplorerService,
+    DelegatesExplorerService,
+    ScalarsExplorerService,
+    GraphQLAstExplorer,
+    GraphQLTypesLoader,
+    GraphQLSchemaBuilder,
+  ],
+  exports: [],
+})
+export class GraphQLFederationModule implements OnModuleInit {
+  private apolloServer: ApolloServerBase;
+
+  constructor(
+    @Optional()
+    private readonly httpAdapterHost: HttpAdapterHost,
+    @Inject(GRAPHQL_MODULE_OPTIONS)
+    private readonly options: GqlModuleOptions,
+    private readonly graphqlFederationFactory: GraphQLFederationFactory,
+    private readonly graphqlTypesLoader: GraphQLTypesLoader,
+    private readonly graphqlFactory: GraphQLFactory,
+    private readonly applicationConfig: ApplicationConfig,
+  ) {}
+
+  static forRoot(options: GqlModuleOptions = {}): DynamicModule {
+    options = mergeDefaults(options);
+
+    return {
+      module: GraphQLFederationModule,
+      providers: [
+        {
+          provide: GRAPHQL_MODULE_OPTIONS,
+          useValue: options,
+        },
+      ],
+    };
+  }
+
+  static forRootAsync(options: GqlModuleAsyncOptions): DynamicModule {
+    return {
+      module: GraphQLFederationModule,
+      imports: options.imports,
+      providers: [
+        ...this.createAsyncProviders(options),
+        {
+          provide: GRAPHQL_MODULE_ID,
+          useValue: generateString(),
+        },
+      ],
+    };
+  }
+
+  private static createAsyncProviders(options: GqlModuleAsyncOptions): Provider[] {
+    if (options.useExisting || options.useFactory) {
+      return [this.createAsyncOptionsProvider(options)];
+    }
+
+    return [
+      this.createAsyncOptionsProvider(options),
+      {
+        provide: options.useClass,
+        useClass: options.useClass,
+      },
+    ];
+  }
+
+  private static createAsyncOptionsProvider(options: GqlModuleAsyncOptions): Provider {
+    if (options.useFactory) {
+      return {
+        provide: GRAPHQL_MODULE_OPTIONS,
+        useFactory: options.useFactory,
+        inject: options.inject || [],
+      };
+    }
+
+    return {
+      provide: GRAPHQL_MODULE_OPTIONS,
+      useFactory: (optionsFactory: GqlOptionsFactory) => optionsFactory.createGqlOptions(),
+      inject: [options.useExisting || options.useClass],
+    };
+  }
+
+  async onModuleInit() {
+    if (!this.httpAdapterHost || !this.httpAdapterHost.httpAdapter) {
+      return;
+    }
+
+    const { printSchema } = loadPackage('@apollo/federation', 'ApolloFederation');
+
+    const { typePaths } = this.options;
+    const typeDefs = await this.graphqlTypesLoader.getTypesFromPaths(typePaths);
+    const mergedTypeDefs = extend(typeDefs, this.options.typeDefs);
```

Comment:
Hello @tuxmachine
Why do you use here getTypesFromPaths instead of mergeTypesByPaths, as it is the case in graphql.module.ts?

Reply:
I think because of this, when we try to create an app with multiple resolvers and gql files, it creates errors of the type `GraphQLError: There can be only one type named "Query"` since it no longer merges these types together.
This is breaking compared to the behavior of graphql.module.ts and the Nest.js documentation where we are allowed to declare the same Query and Mutation types several times

Reply:
Are you using `extend type Query` ? Without the `extend` I get the same error. 

I assumed @marcus-sa had a good reason not to use the `mergeTypesByPaths`, I'll look into it.

Reply:
You need to use `extend type` otherwise it won't work (this is also documented in Apollo Federation & GraphQL)

I remember I had a good reason for not using it, but it's been so long that I've forgot.

Reply:
Thanks for pitching in :) I'll have a look 

The `mergeTypesByPaths` uses an external library, so I can imagine it didn't support federation back then

Reply:
`mergeTypesByPaths` resolves to nearly an identical schema (just adds a `schema` property) so the issue seems to have been resolved.

Reply:
seems great, thanks @tuxmachine 💚